File: python/programmers/Graph/순위.py
def solution(n, results):
    answer = 0
    win,lose={},{}
    for i in range(1,n+1):
        win[i],lose[i]=set(),set()
    for i,j in results:
        win[i].add(j)
        lose[j].add(i)
    for i in range(1,n+1):
        for j in lose[i]:
            win[j].update(win[i])
        for j in win[i]:
            lose[j].update(lose[i])
    for i in range(1,n+1):
        if len(win[i])+len(lose[i])==n-1:
            answer+=1
    return answer

File: javascript/programmers/Level3/순위.js
function solution(n, results) {
  var answer = 0;
  const win = [];
  const lose = [];
  const cnts = new Array(n + 1).fill(0);
  for (let i = 0; i < n + 1; i++) {
    win.push([]);
    lose.push([]);
  }
  results.forEach((result) => {
    win[result[0]].push(result[1]);
    lose[result[1]].push(result[0]);
  });
  console.log(win);
  console.log(lose);
  const check = (adj, cur) => {
    const visited = new Array(n + 1).fill(false);
    const q = [cur];
    while (q.length > 0) {
      const x = q.shift();
      if (visited[x]) continue;
      visited[x] = true;
      cnts[cur]++;
      adj[x].forEach((nx) => {
        if (!visited[nx]) q.push(nx);
      });
    }
  };
  for (let i = 1; i < n + 1; i++) {
    check(win, i);
    check(lose, i);
    if (cnts[i] === n + 1) answer++;
  }
  return answer;
}
